tools/release_builder.py

```diff
@@ -74,33 +74,81 @@ def _run_cmd(self, cmd, exit_on_fail=True, dry_run_return = 0):
-                badfile, self._stub_universe_url, local_zip_path))
-        zipin.extractall(scratchdir)
-        # check for stub-universe-pkgname/repo/packages/P/pkgname/0/:
-        pkgdir_path = os.path.join(
-            scratchdir,
-            'stub-universe-{}'.format(self._pkg_name),
-            'repo',
-            'packages',
-            self._pkg_name[0].upper(),
-            self._pkg_name,
-            '0')
-        if not os.path.isdir(pkgdir_path):
-            raise Exception('Didn\'t find expected path {} after unzipping {}'.format(
-                pkgdir_path, local_zip_path))
-        os.unlink(local_zip_path)
-        return pkgdir_path
+        '''Returns the path to the package directory in the stub universe.'''
+        stub_universe_file = urllib.request.urlopen(self._stub_universe_url)
+
+        stub_universe_extension = os.path.splitext(self._stub_universe_url)[1]
+        if stub_universe_extension == '.zip':
+            # stub universe zip package (universe 2.x only)
+            zipin = zipfile.ZipFile(stub_universe_file, 'r')
+            badfile = zipin.testzip()
+            if badfile:
+                raise Exception('Failed to unpack {} in downloaded {}'.format(
+                    badfile, self._stub_universe_url))
+            zipin.extractall(scratchdir)
+            # check for (and return) path to stub-universe-pkgname/repo/packages/P/pkgname/0/:
+            pkgdir_path = os.path.join(
+                scratchdir,
+                'stub-universe-{}'.format(self._pkg_name),
+                'repo',
+                'packages',
+                self._pkg_name[0].upper(),
+                self._pkg_name,
+                '0')
+            if not os.path.isdir(pkgdir_path):
+                raise Exception('Didn\'t find expected path {} after unzipping {}'.format(
+                    pkgdir_path, self._stub_universe_url))
+            return pkgdir_path
+        elif stub_universe_extension == '.json':
+            # stub universe json file (universe 3.x+ only)
+            stub_universe_json = json.loads(stub_universe_file.read().decode('utf-8'), object_pairs_hook=OrderedDict)
+
+            # put package files into a subdir of scratchdir: avoids conflicts with reuse of scratchdir elsewhere
+            pkgdir = os.path.join(scratchdir, 'stub-universe-{}'.format(self._pkg_name))
+            os.makedirs(pkgdir)
+
+            if 'packages' not in stub_universe_json:
+                raise Exception('Expected "packages" key in root of stub universe JSON: {}'.format(
+                    self._stub_universe_url))
+            if len(stub_universe_json['packages']) != 1:
+                raise Exception('Expected single "packages" entry in stub universe JSON: {}'.format(
+                    self._stub_universe_url))
+
+            package_json = stub_universe_json['packages'][0]
+
+            def extract_json_file(package_dict, name):
+                file_dict = package_dict.get(name, None)
```

Comment:
Odd;  `dict.get(key)` is the same as `dict.get(key, None)`

I'm not sure yet whey you're processing package_dict destructively, but I'd probably have just written

```python
if name in package_dict:
    file_dict = package_dict.pop(name)
    ....
```